--- package/boot/grub2/Makefile	2020-02-12 14:29:15.000000000 +0800
+++ OpenWrt-UEFI-Support/src/package/boot/grub2/Makefile	2020-02-13 16:03:56.188091571 +0800
@@ -25,10 +25,14 @@
 
 PKG_FLAGS:=nonshared
 
+#PATCH_DIR := ../patches
+#HOST_PATCH_DIR := ../patches
+#HOST_BUILD_DIR := $(BUILD_DIR_HOST)/$(PKG_NAME)-$(PKG_VERSION)
+
 include $(INCLUDE_DIR)/host-build.mk
 include $(INCLUDE_DIR)/package.mk
 
-define Package/grub2
+define Package/grub2/Default
   CATEGORY:=Boot Loaders
   SECTION:=boot
   TITLE:=GRand Unified Bootloader
@@ -36,18 +40,6 @@
   DEPENDS:=@TARGET_x86||TARGET_x86_64
 endef
 
-define Package/grub2-editenv
-  CATEGORY:=Utilities
-  SECTION:=utils
-  SUBMENU:=Boot Loaders
-  TITLE:=Grub2 Environment editor
-  URL:=http://www.gnu.org/software/grub/
-  DEPENDS:=@TARGET_x86||TARGET_x86_64
-endef
-
-define Package/grub2-editenv/description
-	Edit grub2 environment files.
-endef
 
 HOST_BUILD_PREFIX := $(STAGING_DIR_HOST)
 
@@ -82,12 +74,3 @@
 	$(SED) 's,(RANLIB),(TARGET_RANLIB),' $(HOST_BUILD_DIR)/grub-core/Makefile.in
 	$(Host/Configure/Default)
 endef
-
-define Package/grub2-editenv/install
-	$(INSTALL_DIR) $(1)/usr/sbin
-	$(INSTALL_BIN) $(PKG_BUILD_DIR)/grub-editenv $(1)/usr/sbin/
-endef
-
-$(eval $(call HostBuild))
-$(eval $(call BuildPackage,grub2))
-$(eval $(call BuildPackage,grub2-editenv))
