--- package/base-files/files/sbin/sysupgrade	2020-02-13 19:28:09.766212177 +0800
+++ OpenWrt-UEFI-Support/src/package/base-files/files/sbin/sysupgrade	2020-02-13 21:30:56.937508092 +0800
@@ -191,6 +191,7 @@
 
 	return 0
 }
+sysupgrade_image_check="platform_check_image"
 
 if [ $SAVE_OVERLAY = 1 ]; then
 	[ ! -d /overlay/upper/etc ] && {
@@ -314,11 +315,17 @@
 		;;
 esac
 
-json_load "$(/usr/libexec/validate_firmware_image "$IMAGE")" || {
-	echo "Failed to check image"
-	exit 1
-}
-json_get_var valid "valid"
+if [ "$(get_magic_long)" == "16faaf61" ]; then
+	platform_check_image()
+	valid="1"
+else
+	json_load "$(/usr/libexec/validate_firmware_image "$IMAGE")" || {
+		echo "Failed to check image"
+		exit 1
+	}
+	json_get_var valid "valid"
+fi
+
 [ "$valid" -eq 0 ] && {
 	if [ $FORCE -eq 1 ]; then
 		echo "Image check failed but --force given - will update anyway!" >&2
